### ISSUE TITLE
About siamese predition threshold

### ISSUE THREAD

**Issue by jinbiaoyu** [avilash/tf_siamese#1]
Hello, I'm a newer of siamese net work. I have a problem do not understand. Why the threshold for prediction is half of training margin? I'm using the siamese for face recognition, and I set the thresold for 
 prediction is 1. and it is same with trainning marin, but i can only get 50% accuracy for test, I don't understand why, can you help me?

**Comment by avilash** [comment]
If you look at the loss function, it tries to push the distances of dissimilar objects towards the margin until it crosses it and distances of similar objects as close to 0.
Thus the networks ends up learning to predict the distances of similar objects as close to 0 and that of dissimilar objects as close to and preferably greater than the margin.
That is the reason behind choosing the threshold as half of the margin.

I assume you have 50% of similar and  50% dissimilar objects in your test data.
It seems all your distance predictions lie on one side of the margin that you set (mostly below the margin) which gives you a prediction of 50% as all them are predicted as similar or dissimilar.

You could debug by plotting the distribution of predictions for similar and dissimilar objects.
If they both are separated apart quite nicely , your network works well and from the distribution you could choose a threshold which suits you. Ideally half of the margin works

**Comment by JinbiaoYu** [comment]
@avilash Well, thank you very much! As you can see, there are two categories int test data, each of them have 50%. The data for train and test are comes from two dataset. So all the test data pair distance are precdicted below the margin.

For the sake of simplicity, I use 10% of the training set as the test set. Taking your suggestion, we plotted the distance distribution of the image pair. With the increase of iteration times, the positive sample distance becomes smaller and the negative sample distance becomes larger. But what amazed me was that the result was the same as before. I guess there should be a problem with my code, but I have not detected any problems there.

In addition, I have calculated three performance indicators for face verification, accuracy, false acceptance rate, false rejection rate, I'm wonder is am right?

```
accuracy = tf.reduce_mean(tf.cast(tf.equal(tf.less(d,m), tf.equal(y, 1)), dtype=tf.float32))
far = tf.reduce_mean(tf.cast(tf.less(d, m), tf.float32) * tf.cast(tf.equal(y, 0), tf.float32))
frr = tf.reduce_mean(tf.cast(tf.greater(d, m), tf.float32) * tf.cast(tf.equal(y, 0), tf.float32))
```
